daemon/test/test_util.c

```diff
@@ -0,0 +1,315 @@
+    g_assert_cmpstr(result, ==, PERMISSIONS_DIRECTORY "/Test.permission");
+    g_free(result);
+}
+
+/* ========================================================================= *
+ * CHANGE TESTS
+ * ========================================================================= */
+
+void test_change_uid()
+{
+    uid_t uid = 0;
+    g_assert_false(change_uid(&uid, 0));
+    g_assert_cmpint(uid, ==, 0);
+    g_assert_true(change_uid(&uid, 1));
+    g_assert_cmpint(uid, ==, 1);
+}
+
+void test_change_boolean()
+{
+    bool boolean = false;
+    g_assert_false(change_boolean(&boolean, false));
+    g_assert_false(boolean);
+    g_assert_true(change_boolean(&boolean, true));
+    g_assert_true(boolean);
+}
+
+void test_change_string()
+{
+    gchar *string = g_strdup("foo"); // consumed
+    g_assert_false(change_string(&string, "foo"));
+    g_assert_cmpstr(string, ==, "foo");
+    g_assert_true(change_string(&string, "bar"));
+    g_assert_cmpstr(string, ==, "bar");
+    g_assert_true(change_string(&string, NULL));
+    g_assert_null(string);
+    g_assert_false(change_string(&string, NULL));
+    g_assert_null(string);
+}
+
+void test_change_string_steal()
+{
+    gchar *string = g_strdup("foo"); // consumed
+    {
+        gchar *tmp = string;
+        gchar *another = g_strdup("foo"); // consumed
+        g_assert_false(change_string_steal(&string, another));
+        g_assert_cmpstr(string, ==, "foo");
+        g_assert_true(string == tmp);
+    }
+    {
+        gchar *another = g_strdup("bar"); // consumed
+        g_assert_true(change_string_steal(&string, another));
+        g_assert_cmpstr(string, ==, "bar");
+        g_assert_true(string == another);
+    }
+    g_assert_true(change_string_steal(&string, NULL));
+    g_assert_null(string);
+    g_assert_false(change_string_steal(&string, NULL));
+    g_assert_null(string);
+}
+
+/* ========================================================================= *
+ * KEYFILE TESTS
+ * ========================================================================= */
+
+void test_keyfile_merge()
+{
+    GKeyFile *keyfile = g_key_file_new();
+
+    {
+        g_assert_true(keyfile_merge(keyfile, TESTDATADIR "/keyfile1.ini"));
+        gchar *value = keyfile_get_string(keyfile, "GroupB", "KeyB1", "default");
+        g_assert_cmpstr(value, ==, "ValueB1");
+        g_free(value);
+    }
+    {
+        g_assert_true(keyfile_merge(keyfile, TESTDATADIR "/keyfile2.ini"));
+        gchar *value = keyfile_get_string(keyfile, "GroupB", "KeyB1", "default");
+        g_assert_cmpstr(value, ==, "OverrideB1");
+        g_free(value);
+    }
+
+    g_key_file_free(keyfile);
+}
+
+void test_keyfile_save()
+{
+    GKeyFile *keyfile = g_key_file_new();
+    keyfile_set_boolean(keyfile, "Foo", "boolean", true);
+    keyfile_set_integer(keyfile, "Foo", "number", 100);
+    g_assert_true(keyfile_save(keyfile, SHAREDSTATEDIR "/test.ini"));
+    g_assert_true(g_file_test(SHAREDSTATEDIR "/test.ini", G_FILE_TEST_EXISTS));
```

Comment:
Should the content be verified as well?

Reply:
Added that. Although I kind of expect that glib does its job when the file exists but I suppose it's still important to verify to some degree.